specs/granite/header_accumulator.md

```diff
@@ -0,0 +1,302 @@
+# Granite - Header Accumulator
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Overview](#overview)
+- [Timestamp Activation](#timestamp-activation)
+- [Rationale](#rationale)
+- [Constants & Definitions](#constants--definitions)
+- [Accumulator Tree Construction](#accumulator-tree-construction)
+  - [Interior Commitment Truncation Rules](#interior-commitment-truncation-rules)
+- [Block Execution Changes](#block-execution-changes)
+  - [Block Validity Changes](#block-validity-changes)
+  - [Header Changes](#header-changes)
+    - [`extraData` format](#extradata-format)
+    - [Header RLP Size Considerations](#header-rlp-size-considerations)
+  - [Accumulator Tree Functions](#accumulator-tree-functions)
+    - [Zero Hashes](#zero-hashes)
+    - [Tree Root](#tree-root)
+    - [Insertion](#insertion)
+    - [Verification](#verification)
+- [Eventual Migration to EIP-7685](#eventual-migration-to-eip-7685)
+- [Forwards Compatibility Considerations](#forwards-compatibility-considerations)
+- [Client Implementation Considerations](#client-implementation-considerations)
+- [Infrastructure Considerations](#infrastructure-considerations)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+The Granite hardfork introduces a new phase of block execution, header hash accumulation. After fork activation,
+every `HEADER_BATCH_SIZE` blocks, the execution layer must append a binary merkle tree of depth
+`HEADER_BATCH_TREE_DEPTH` containing the previous `HEADER_BATCH_SIZE` block hashes (exclusive of the current block) to
+the header accumulation merkle tree of depth `ACCUMULATOR_TREE_DEPTH` in the previous header accumulation block.
+
+## Timestamp Activation
+
+Granite, like other OP Stack network upgrades, is activated at a timestamp. Changes to the L2 Block execution rules
+outlined in this specification are applied when the `L2 Timestamp >= activation time`.
+
+## Rationale
+
+After the activation of the [interop hardfork](../interop/overview.md), the computational complexity of proving the
+OP Stack's state transition will increase. The current plan for proving the inclusion of remote logs within chains
```

Comment:
I am not sure either, or at least not sure enough to give it a concrete time-complexity.

From what I do know, it will multiply, with a minima of linear complexity increase (scaling with # of chains in the dependency set). Though, due to us having to reproduce the L2 block for _every block that a remote log is included in_, the computational complexity also grows with the number of relayed messages in the current frame (note: I'm using "frame" here as a non-official term; What I mean is the "frame" in which we're resolving cross-chain dependencies.)

Reply:
The maximum we have to execute in a single FP VM run is a single L2 block (from a single chain). We can extend the output root bisection game to avoid having to execute a block from every L2 chain in the set within a single VM execution. So the cost to having additional chains is pretty negligible. The issue is actually just how old a log can be referenced by an executing message because we have to walk back one block at a time (by hash) to retrieve that referenced log which gets expensive if it's a long way in the past.